code/controllers/subsystems/lighting.dm

```diff
@@ -100,59 +110,88 @@ var/datum/controller/subsystem/lighting/SSlighting
 	var/list/curr_corners = corner_queue
 	var/list/curr_overlays = overlay_queue
 
-	while (curr_lights.len)
-		var/datum/light_source/L = curr_lights[curr_lights.len]
-		curr_lights.len--
+	while (curr_lights.len && lq_idex <= curr_lights.len)
+		var/datum/light_source/L = curr_lights[lq_idex++]
 
-		L.update_corners()
+		if (L.needs_update != LIGHTING_NO_UPDATE)
+			L.update_corners()
 
-		L.needs_update = LIGHTING_NO_UPDATE
+			L.needs_update = LIGHTING_NO_UPDATE
 
-		processed_lights++
+			processed_lights++
 
 		if (no_mc_tick)
 			CHECK_TICK
 		else if (MC_TICK_CHECK)
 			break
 
+	if (lq_idex > 1)
+		curr_lights.Cut(1, lq_idex)
```

Comment:
Does this break if curr_lights.len == 0?

Reply:
This line shouldn't be reached if the above loop doesn't run, as `lq_idex` will still be 1.